### PR TITLE
Trim leading whitespace when uploading from url

### PR DESCRIPTION
**Pull request by jdavcs**
Solves this: https://github.com/galaxyproject/galaxy/issues/8196

### CODE REVIEW

File: lib/galaxy/tools/actions/upload_common.py [hunk 1]
```diff
@@ -120,11 +120,14 @@ def persist_uploads(params, trans):
                                                    local_filename=local_filename)
             elif type(f) == dict and 'local_filename' not in f:
                 raise Exception('Uploaded file was encoded in a way not understood by Galaxy.')
-            if 'url_paste' in upload_dataset and upload_dataset['url_paste'] and upload_dataset['url_paste'].strip() != '':
-                upload_dataset['url_paste'] = datatypes.sniff.stream_to_file(
-                    StringIO(validate_url(upload_dataset['url_paste'], trans.app.config.fetch_url_whitelist_ips)),
-                    prefix="strio_url_paste_"
-                )
+
+            if 'url_paste' in upload_dataset and upload_dataset['url_paste']:
```
Comment (martenson) [inline lib/galaxy/tools/actions/upload_common.py]:
```suggestion
            if 'url_paste' in upload_dataset:
```

Comment (jdavcs) [reply lib/galaxy/tools/actions/upload_common.py]: I think we need the second condition: 'url_paste' in dict tests the presence of the key; the value still may be None - in which case calling strip() on the value will raise an attributeerror, no?

File: lib/galaxy/tools/actions/upload_common.py [hunk 2]
```diff
@@ -120,11 +120,14 @@ def persist_uploads(params, trans):
                                                    local_filename=local_filename)
             elif type(f) == dict and 'local_filename' not in f:
                 raise Exception('Uploaded file was encoded in a way not understood by Galaxy.')
-            if 'url_paste' in upload_dataset and upload_dataset['url_paste'] and upload_dataset['url_paste'].strip() != '':
-                upload_dataset['url_paste'] = datatypes.sniff.stream_to_file(
-                    StringIO(validate_url(upload_dataset['url_paste'], trans.app.config.fetch_url_whitelist_ips)),
```
Comment (martenson) [inline lib/galaxy/tools/actions/upload_common.py]: Maybe just adding a `strip()` to this line instead could work. So the `else` branch is triggered the same as before

File: lib/galaxy/tools/actions/upload_common.py (hunk 1 above)
Comment (martenson) [reply lib/galaxy/tools/actions/upload_common.py]: True, could be the case of None.

File: lib/galaxy/tools/actions/upload_common.py (hunk 2 above)
Comment (jdavcs) [reply lib/galaxy/tools/actions/upload_common.py]: We can, of course. Very minor issue though, IMHO: we're calling strip() twice. It doesn't matter for performance or readability, but if down the road we need to do something else to that value, we'll have to remember to call strip (a third time). If the pasted url should have no whitespace, I'd trim it before doing any kind of processing. But that's a very minor consideration; besides, there's the added line as a downside.. Up to you.

Comment (martenson) [reply lib/galaxy/tools/actions/upload_common.py]: In your current implementation if you get `"    "` you return `""`, in the previous one you'll return `None`. Not sure how important it is in this case but upload can to be pretty finicky.

Comment (jdavcs) [reply lib/galaxy/tools/actions/upload_common.py]: sounds very important! checking....

Comment (jdavcs) [reply lib/galaxy/tools/actions/upload_common.py]: @martenson what do you think of this approach? I split it in two: (a) getting the value if it exists, and (b) processing that value. I think this is slightly more readable, but I'm torn. If you think it's too much, let's go with what you suggested - just calling strip() again.
But if this does work, i'll add a couple unit tests.

Comment (martenson) [reply lib/galaxy/tools/actions/upload_common.py]: Are you still sure you'd choose it over this?

```diff
-                    StringIO(validate_url(upload_dataset['url_paste'], trans.app.config.fetch_url_whitelist_ips)),
+                    StringIO(validate_url(upload_dataset['url_paste'].strip(), trans.app.config.fetch_url_whitelist_ips)),
```

Comment (martenson) [reply lib/galaxy/tools/actions/upload_common.py]: > if down the road we need to do something else to that value

That url value disappears in either case, it is either None, or a file handle.

Comment (jdavcs) [reply lib/galaxy/tools/actions/upload_common.py]: It does disappear.. And there's really no place we would reuse that function in this file. So yes, I think you're right, let's keep it simple. Will submit in a min.

Comment (jdavcs) [reply lib/galaxy/tools/actions/upload_common.py]: Done.

**Comment by nsoranzo** [comment]
2 API tests are failing.

**Comment by nsoranzo** [comment]
I fear that's not the correct approach, what if the user wants to upload some text (not a list of URLs) which starts on purpose with some spacing?

**Comment by jdavcs** [comment]
@nsoranzo - yes. But then, I'm wondering, wouldn't it it be better to treat data (to upload) and URLs (to fetch data from) differently? Those are very different processes, and I think the system should distinguish between the two as early as possible. How to process the CR character should be a consideration for the pasted or fetched data, but, I think, it should have no bearing on how we treat a list of URLs. In other words, first get the data (pasted input or fetch from pasted URLs), and only then process it. 
So, what if we decouple fetching data from processing data? As one of the benefits, trimming surrounding whitespace or doing any other URL-related processing would become so easy.

**Comment by martenson** [comment]
Could be enough to move the lstrip to the `validate_url` method.

**Comment by jdavcs** [comment]
@martenson - yep, will do right away - should solve the immediate problem, but I'm curious what you think about my other suggestion.

File: lib/galaxy/tools/actions/upload_common.py [hunk 3]
```diff
@@ -26,6 +26,8 @@ def validate_url(url, ip_whitelist):
     if not (url.lstrip().startswith('http://') or url.lstrip().startswith('https://')):
         return url
 
+    # Strip leading whitespace
+    url = url.lstrip()
```
Comment (nuwang) [inline lib/galaxy/tools/actions/upload_common.py:30]: should this be just `url = url.strip()`, since whitespace on either side is invalid?

**Comment by jdavcs** [comment]
@nuwang I try to minimize the footprint here. Stripping on both ends in a previous iteration led to 2 api tests failing (when testing for POSIX standard on/off): the carriage return character ended up being stripped off, while the test expected it to be preserved. This may or may not be relevant for this iteration, since we modify a local variable.. But I didn't want any surprises, so I changed the minimum.
